Filament loading state fixes

wavexx opened the pull request:
There are some issues in the management of "filament autoloading" when the printer is idle. Because the autoloading is performed asynchronously, we need to ensure incompatible actions are disabled while loading is taking place.

With this PR we make impossible to perform another autoload/unload/filament action when one is already taking place (the autoload/load/unload menus are now hidden while autoloading is taking place).

To do so we check for the loading_flag. This was previously done for loading only, but doing so for unload makes the behavior uniform (#2318 performs more than a single move and is also non-blocking - you shouldn't load while the old unload is still pending).

Using bFilamentState flag would have been a little bit cleaner here (removing the need for loading_flag), but it's currently used both as a flag for autoload and also for the autoload state...

We also now prevent to start an SD print while autoload is taking place (fixes https://github.com/prusa3d/Prusa-Firmware/issues/2651). You can still navigate the SD card when waiting, but clicking on the file will just beep (see my comments on b0561a0aa1ba757b6e442c17995f17fc4e969d2c)

As a side note, in the first commit, we change PRINTER_ACTIVE to consider any custom command to be non-idle, which is logical. All custom commands are incompatible with each other, and will return the printer to Idle when complete (LongPause included). Checking for Layer1Cal in the main menu is simply incorrect.

Change in memory (MK3S+ Multilang):
Flash: -8 bytes
SRAM: -1 byte

wavexx committed:
Consider any lcd_custom_command to be "ACTIVE"

All custom commands are transitory and eventually switch back to Idle
state by themselves.

It doesn't make any sense to explicitly check for Layer1Cal: any
non-idle state is active by design.

Fix this check in the main menu. This is probably incomplete (Layer1Cal
is incorrectly used in several other places).

wavexx committed:
Set/clear loading_flag also during unload

wavexx committed:
Do not allow filament actions while an existing one is taking place

wavexx committed:
Refuse to start a SD print until filament loading is complete

Do not blidnly clear the loading_flag, check for it!

Just disallowing the SD menu while loading is being performed is not
sufficient, since the menu can be entered also by inserting card while
loading is taking place.

This is also nicer in behavior, as we allow to navigate the SD card
while loading.

gudnimg commented:
Is `loading_flag` only applicable for autoloading? Should we consider renaming it to `auto_loading_flag` or something to that effect?

wavexx commented:
> Is loading_flag only applicable for autoloading? Should we consider renaming it to auto_loading_flag or something to that effect?

I would be in favor to rename this as auto[_]loading_flag, but we need to this after the MK3_MMU2 branch is merged.

wavexx committed:
Consistently replace loading_flag with eFilamentAction

Remove loading_flag and check for eFilamentAction instead which already
flags both load/unload (in addition to mmu actions).

Correctly transition from AutoLoad to Load as soon as the operation
cannot be cancelled anymore as opposed to resetting it.

wavexx committed:
Improve FilamentAction comments

wavexx commented:
@gudnimg rereading the code it seems that now eFilamentAction does what we need, so I removed loading_flag completely.

The old code was resetting the "autoload" flag for some reason. I changed the meaning so that the "AutoLoad" state is the moment when autoload has been triggered, but loading is still not taking place. AutoLoad can be cancelled (if you're still in the preheat menu), while "Load" is already starting to feed the filament and cannot be undone.

IMHO this is a lot saner, and I hope I understood all existing scenarios.

Can you review the MMU part? I added one case in lcd_wizard_load which I think was missing.

@3d-gussner because eFilamentAction is now used everywhere this requires some retesting of #2651 but also retesting all scenarios with loading/unloading, autoloading+cancel, and using the menu load/load filament correctly clears the load flag, otherwise you won't see the "load/unload" menus anymore.

gudnimg reviewed (approved):
Looks good to me. I think you're right about `eFilamentAction = FilamentAction::MmuLoad` being missing before.

wavexx commented:
Should fix #3297

gudnimg commented:
@wavexx I rebased the PR to resolve the merge conflicts. Hopefully I didn't break anything. The PR was quite a bit behind MK3 branch :)

gudnimg commented:
Rebased again. For some reason I had to re-do the entire rebase (~600 commits)

3d-gussner reviewed (approved):
Approved and fixes issues like https://github.com/prusa3d/Prusa-Firmware/issues/3595
Other tests with MK404 show that this PR is an improvement to previous version.
